sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+kernel's requirements, and it must raise an `errc::kernel_not_supported`
+exception if it does not.
+
+When a kernel is submitted to a device, the runtime finds the device image that
+contains the kernel and also finds the kernel's entry in the
+"SYCL/requirements" property set.  This entry tells the set of requirements for
+the kernel.  If the target device does not support all of these requirements,
+then the runtime raises `errc::kernel_not_supported`.  This check can be done
+before the device image is JIT compiled, so the exception can be thrown
+synchronously.
+
+If the kernel imports device function symbols from a shared library as defined
+in [Device Code Dynamic Linking][7], the runtime first identifies all the
+device images that define these exported device functions.  Before attempting
+to link them together, the runtime finds the entries for the exported device
+functions in their "SYCL/requirements" property sets and checks that the device
+supports all these requirements.  If it does not, the runtime throws
+`errc::kernel_not_supported`.
+
+Whenever the runtime submits a SPIR-V image to the backend for online
+compilation, it must do the following additional steps:
+
+* Compute the union of all requirements from all entries in the image's
+  "SYCL/requirements" property set.
+
+* Query the target device to see whether it supports each of these
+  requirements, yielding either `True` or `False` for each one.
+
+* Set the value of each associated specialization constant when JIT compiling
+  the device image for this target device.
+
+Note that the runtime's cache of compiled device images does not need any
+special modification because the cache already needs to know the values of all
+the specialization constants that were used to compile the device image.  We
+just need to make sure the cache is also aware of the specialization constants
+which correspond to the requirements from the "SYCL/requirements" property set.
+
+#### Modifications to the GEN compiler
+
+The GEN compiler, of course, needs to be modified to implement the new
+`ConditionalINTEL` SPIR-V decoration.  It must discard any function with this
+decoration (unless the corresponding specialization constant is `True`), and it
+must not raise any sort of error due to compilation of these discarded
+functions.
+
+### AOT case
+
+The AOT case uses exactly the same solution as the JIT case described above,
+but there is one extra steps.  For the AOT case, the post-link tool must set
+the values of the specialization constants that correspond to the requirements
+for the kernel or exported device function, using the device named in the
+"-fsycl-targets" command line option.  After doing this, the post-link tool
+calls the AOT compiler to generate native code from SPIR-V as it normally does.
+If more than one target device is specified, the post-link tool sets the
+specialization constants separately for each device before generating native
+code for that device.
```

Comment:
A lot of questions here:

`-fsycl-targets` at the moment only contains triple and it can't properly point us to a particular device. For example different HW generations could have different HW capabilities and therefore different list of supported aspects, but triple is the same. For GPU we select a device by passing extra arguments to the backend, i.e. we invoke `ocloc` with `-device [device-name]` option.

`sycl-post-link` does not call `ocloc` - this is a driver responsibility.

`sycl-post-link` does not set values of specialization constants - the only existing interface is through calling corresponding L0/OpenCL API or generating SPIR-V with a proper default value from the beginning.

This is related to my concern about AOT in one of my comments above: another idea here would be driver to launch AOT compiler _before_ invoking `sycl-post-link` to gather device capabilities in some predefined machine-readable format and then pass this info into `sycl-post-link` so it can generate correct defaults for spec constants in the resulting SPIR-V file.

Reply:
Thanks, this helps a lot with my understanding of the AOT flow.

> another idea here would be driver to launch AOT compiler before invoking sycl-post-link to gather device capabilities in some predefined machine-readable format and then pass this info into sycl-post-link so it can generate correct defaults for spec constants in the resulting SPIR-V file.

This does not seem ideal because we would need to run sycl-post-link multiple times in the case when we AOT compile for more than one device.  (Because we would need to set different default values for the spec constants depending on the device.)

Instead, what do you think about this design:

* It seems like `ocloc` does not currently have a command line option that can be used to set the value of a specialization constant when compiling a SPIR-V module.  This seems like functionality that would be generally useful, so let's add such an option.  Usage would be like `-spec_const <id>=<value>`, where `<id>` is the numeric SpecID identifier for a SPIR-V specialization constant and `<value>` is the requested value.  Ideally, we'd support Boolean, integer, and floating point values, but for this design we need only Boolean.

* We need some way for the `sycl-post-link` tool to pass information back to the driver about the specialization constants in the SPIR-V module(s).  We choose one of the following depending on what's easier:
  - The `sycl-post-link` tool already generates this information in the "SYCL/requirements" property set.  Is the content of this property set already sent back to the driver for packaging with `clang-offload-wrapper`?  If so, the driver can just read the content of this property set directly.
  - Alternatively, `sycl-post-link` can generate a meta-data file in some format that is convenient for the driver, and the driver can read the information from this meta-data file.

* You say that the driver already passes `-device [device-name]` options to `ocloc`, so it must somehow be able to tell which device(s) we are compiling for.  Therefore, the driver can also be responsible for determine whether each `aspect` is true or false for each device.  The runtime also needs to know this information, so we will store the information in some common file in the source tree, which is shared by both the driver and the runtime.

* The driver uses information from the meta-data file (from `sycl-post-link`) to learn which spec constants are contained by the SPIR-V module and also uses this information to learn the associated `aspect`.  The driver then passes options to `ocloc` to set the spec constant `True` / `False` as appropriate for that device.

Reply:
The scheme you have suggested looks similar to what I was talking about here: https://github.com/intel/llvm/pull/3781#discussion_r639199999.

A few notes:

> We need some way for the sycl-post-link tool to pass information back to the driver about the specialization constants in the SPIR-V module(s). We choose one of the following depending on what's easier:

AFAIK, driver works in the following manner: it fully builds a sequence of commands to execute and only then launches them. Commands can't be conditional and we can't change the list during its execution. @mdtoguchi and @AGindinson can provide more details here. Instead, we should be talking about propagating some info from one tool to another through input/output files of those tools. The phrase "pass information back to the driver" should be "generate a file, which is later passed to AOT compiler" or something like this, per my understanding.

> The `sycl-post-link` tool already generates this information in the "SYCL/requirements" property set. Is the content of this property set already sent back to the driver for packaging with clang-offload-wrapper? If so, the driver can just read the content of this property set directly.

Info is stored in a file, driver just ensures that the file is passed to `clang-offload-wrapper` without reading its content.

> You say that the driver already passes -device [device-name] options to ocloc, so it must somehow be able to tell which device(s) we are compiling for. 

I would like @mdtoguchi and/or @AGindinson to comment here, but my current understanding is that we don't have a generic scheme for selecting HW generation/type, related feature requests are #1382 and #1170. The problem is that currently we select device by specifying target backend options, which are unique for each backend and from the driver point of view it is an arbitrary string which is just passed as-is to the tool: it is not analyzed/inspected in any way by the driver.

> Therefore, the driver can also be responsible for determine whether each aspect is true or false for each device. 

How exactly do you suggest to implement this? Will it be hardcoded for each device or will it be queried through runtime call or a separate tool?

> The driver then passes options to ocloc to set the spec constant True / False as appropriate for that device.

Since the list of commands and their arguments must be known prior to their execution, the interface here would likely be a config file or something like this.

Reply:
>> The sycl-post-link tool already generates this information in the "SYCL/requirements" property set. Is the content of this property set already sent back to the driver for packaging with clang-offload-wrapper? If so, the driver can just read the content of this property set directly.
>
> Info is stored in a file, driver just ensures that the file is passed to clang-offload-wrapper without reading its content.
>
> [...]
>
> Since the list of commands and their arguments must be known prior to their execution, the interface here would likely be a config file or something like this.

If we do not want to change the driver to read the file (containing the device image property sets), maybe we could do something like this instead:

* Create a new tool that is invoked by the driver that generates a config file for `ocloc`.
* Driver invokes this new tool, passing the file that contains the device image property sets.
* The new tool uses that information to generate the config file, including the `-spec_const <id>=<value>` parameters I describe above.
* Driver invokes the `ocloc` tool with this generated config file.

> How exactly do you suggest to implement this? Will it be hardcoded for each device or will it be queried through runtime call or a separate tool?

I don't think we can assume that the target device is present on the compilation system, so I don't think we can query the device at compilation time.  I was thinking that we would have a hard-coded table mapping each device to its set of aspects.  This information could also be used by the DPC++ runtime to implement `device::has(aspect)`.

Reply:
> > How exactly do you suggest to implement this? Will it be hardcoded for each device or will it be queried through runtime call or a separate tool?
>
> I don't think we can assume that the target device is present on the compilation system, so I don't think we can query the device at compilation time. I was thinking that we would have a hard-coded table mapping each device to its set of aspects. This information could also be used by the DPC++ runtime to implement `device::has(aspect)`.

Personally, I don't think that hardcoding device capabilities is a good idea, because:
- There are features, which are not natively supported by a particular device, but they might be emulated by JIT compiler/driver and therefore their support depends not on the HW, but on the driver version
- Older compiler won't know about newer devices and therefore the result might be unexpected - this might not be so important for end users, but every time we enable new device, it will require extra work to get some basic (i.e. native to the new device) features working

Reply:
Keep in mind that we are only using the hard-coded device information when we AOT-compile the device code.  This essentially runs the JIT compiler at AOT compilation time.

Are you suggesting that the AOT compiler should dynamically query `ocloc` to find the set of device features it supports?  I suppose this would avoid duplicating some information in both the AOT compiler and in the driver.  The downside is that it adds some more complexity to the design and adds another component that needs to be modified.

Reply:
We had an offline meeting to discuss this (and other things).  We decided to hard-code the set of aspects that each device supports in a textual config file, and we will provide a way for the user to choose an alternate version of this config file.  This design document has already been updated to include this information.